### PR TITLE
Headless SVG rendering

### PR DESCRIPTION
**Pull request by marcprux**
Implements stand-along headless SVG rendering as requested at https://github.com/uwdata/vega/issues/10. Only SVG headless rendering is implemented. I'll work on canvas once this PR is merged.

This PR also adds test cases that render all the example specs using both the headless "xml" renderer and the normal "svg" renderer under JSDOM and compares that the outputs are exactly the same (modulo JSDOM bugs). This means that vega had a bunch more devDependencies, but these test cases exercise such a high swath of vega functionality that I believe it is worth it. It will also be useful to have automatic validation that the example specs continue to render.

### CODE REVIEW

**Comment by arvind** [comment]
Hi Marc, thanks for this. Will look through it more carefully, and merge it, soon. One question I did have is regarding an issue I faced when trying to get this to work myself. I wasn't able to figure out how to get the node requires working alongside the requirejs/AMD format, especially with the requirejs compiler that puts together vega2.js and vega2.min.js. Having canvas excluded and shim'd was fine, but then it choked on require('fs'). Glancing through your commits, I don't think I see any handling for that (and others like http)? I might've missed it, or maybe you've found a better system that works?

**Comment by marcprux** [comment]
Regarding the node modules, I spend a very long time trying to figure out how to get it to work the "correct" way, but I never did. My ugly compromise solution was to expose `define` as an external API from `src/_end.js` so you can inject a node module from the outside world, and then invoke that with "canvas" and "fs" from `bin/vg2svg`

As for the optimizer, I found (through trial and error) that if you call `require` with a variable instead of a static string, then it won't try to resolve the module locally, so I just do `var name = "canvas"; require(name);"` and that gets around the module error.

It's all kind of ugly, but it works in node and in the browser, so it might hold until someone better versed in AMD+requirejs+almond can suggest a proper solution (assuming there is one).

**Comment by arvind** [comment]
Thanks, Marc. Yeah, I wasn't able to get something clean to work either. It might be that we need to use some other module/build system. But, I don't think browserify (the most viable alternative) is much better for this case...

**Comment by marcprux** [comment]
Agreed. I made a new issue for the module improvements: https://github.com/uwdata/vega/issues/22

**Comment by marcprux** [comment]
One outstanding question is the best name for the headless SVG renderer.
1. "xml" (since it generates an SVG XML string)
2. "svg-xml" (and would we re-name the existing in-browser "svg" renderer to "svg-dom"?)
3. "string" (since it emits a string)
4. "headless" (but then what would the eventual headless canvas renderer be called?)
5. "vector"

I picked "xml", despite it seeming kind of vague. It's a minor issue, but since it affects the name of the render/xml/ directory, we may want to agree on a name before merging the PR.

**Comment by jheer** [comment]
I would definitely urge us to keep "svg" in the module name for the sake of
clarity. I'd be happy with "svg-string" or "svg-headless" or something of
that sort.

**Comment by arvind** [comment]
Yup, I prefer "svg-headless" or "headless-svg" -- the latter being more consistent with "jsdom"

**Comment by marcprux** [comment]
OK, I renamed it to "svg-headless". Travis seems happy with it.

File: src/parse/spec.js [hunk 1]
```diff
@@ -28,9 +28,10 @@ define(function(require, exports, module) {
         padding: parsePadding(spec.padding),
         signals: parseSignals(model, spec.signals),
         predicates: parsePredicates(model, spec.predicates),
-        marks: parseMarks(model, spec, width, height),
-        data: parseData(model, spec.data, function() { callback(viewFactory(model)); })
+        marks: parseMarks(model, spec, width, height)
       });
+
+      model.data(parseData(model, spec.data, function() { callback(viewFactory(model)); }))
```
Comment (arvind) [inline src/parse/spec.js]: Marc, could you explain this change to me? This isn't correct usage of model.data, which is a getter/setter for data sources.

Comment (marcprux) [inline src/parse/spec.js]: The general rationale of moving the parseData outside of the defs call is to eliminate the need for the setTimeout() call before invoking the callback.

But I'm not sure what I was thinking w.r.t. the model.data() part. I've removed that.

Comment (arvind) [inline src/parse/spec.js]: Gotcha. Is this (and the similar change made to interactors) necessary to support headless rendering? If not, would you mind reverting these changes and submitting them separately? I would like to keep PRs tightly scoped to each issue to make it easier for me to track all the changes being submitted for consideration. Thanks :)

Comment (marcprux) [inline src/parse/spec.js]: OK, I've reverted the changes. I'll re-introduce them in a new PR.

**Comment by marcprux** [comment]
@arvind When do you think this PR will be merged? I have some other test cases and bug fixes waiting on this one.

**Comment by arvind** [comment]
It will probably be a few weeks until I get around to merging it, because I'm currently working on a set of optimizations to the code base. Not sure why merging the PR is blocking you on other tasks?

**Comment by arvind** [comment]
Hi Marc, I'm finally getting a chance to review your PR. Do you mind doing a pull on v2 and merging the changes we've made to it since you branched. That will make it easier for me to review and merge. Thanks!

**Comment by marcprux** [comment]
Arvind-

Will do. I did a trial run after you last major commit a few days ago and there are enough conflicts that it'll be a bit of work. I'll try again with the recent merges; I should get to it sometime next week.

```
-Marc
```

**Comment by marcprux** [comment]
OK, I've re-implemented the headless rendering in a new PR: https://github.com/uwdata/vega/pull/44

I'm closing this one.

**Pull request closed without merging**